src/appl.js

```diff
@@ -5,25 +5,160 @@ const fs = require('fs');
 const os = require('os');
 
 const logger = require('./logger');
+const filterFactory = require('./filter');
+const utils = require('./utils');
 
 class Appl {
 
   /*
   *
   * params:
   */
-  constructor(verbose, cwd) {
+  constructor(verbose, cwd, cliHome, detach, localRepo) {
     this.logger = logger.create(verbose);
     this.cwd = cwd;
+    this.cliHome = cliHome;
+    this.detach = detach;
+    this.localRepo = localRepo;
+    this.home = this.cwd;
+    this.filter = filterFactory.create(this.logger);
+    this.tln = {};
+
+  }
+
+  async init() {
+    await this.filter.configure();
+    this.tln = Object.freeze({
+      logger: this.logger,
+      isWindows: () => this.filter.isWindows(),
+      isLinux: () => this.filter.isLinux(),
+      isDarwin: () => this.filter.isDarwin(),
+      getOsInfo: () => this.filter.getOsInfo(),
+      unpackId: (id) => {
+        const arr = id.split('-');
+        const version = arr.pop();
+        return {name: arr.join('-'), version};
+      },
+      copyTemplate: (tln, script, src, dest, tail = []) => utils.copyTemplate(tln, script, src, dest, tail),
+      canInstallComponent: (tln, id, home) => utils.canInstallComponent(tln, id, home),
+      getDownloadScriptById: (tln, id, distrs) => utils.getDownloadScriptById(tln, id, distrs),
+      getDownloadScript: (tln, dist) => utils.getDownloadScript(tln, dist)
+    })
+
+    //
+    let folders = [];
+    let detached = false;
+    const tmpPath = path.join(os.tmpdir(), 'tln');
+    // find projects' root and current component
+    if (this.detach || this.localRepo) {
+      // we are in detached mode
+      this.localRepo = this.localRepo || tmpPath;
+      detached = true;
+    } else {
+      // find topmost level folder with with tln descs
+      let p = this.home;
+      let noConfig = !utils.isConfigPresent(p);
+      while (!this.isRootPath(p)) {
+        p = path.dirname(p);
+        if (utils.isConfigPresent(p)) {
+          this.home = p;
+          noConfig = false;
+        }
+      }
+      //
+      // build chain of components from projects home to the current folder
+      const rel = path.relative(this.home, this.cwd);
+      if (rel) {
+        folders = rel.split(path.sep);
+      }
+      //
+      // shared components location
+      if (this.isRootPath(this.cwd) || noConfig) {
+        this.localRepo = tmpPath;
+        detached = true;
+      } else {
+        // at project's root level
+        this.localRepo = this.home;
+      }
+    }
+    fs.mkdirSync(this.localRepo, { recursive: true });
+    this.rootComponent = require('./component').createRoot(this.logger, this.tln, this.localRepo, this.cliHome);
+    this.currentComponent = this.rootComponent;
+    if (detached) {
+      this.currentComponent = await this.rootComponent.createChild(this.cwd);
+    } else {
+      for(const folder of folders) {
+        this.currentComponent = await this.currentComponent.buildChild(folder, true);
+      };
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this empty statement. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://159.65.230.204:9000/coding_rules#rule_key=javascript%3AS1116)